django/contrib/gis/geos/coordseq.py

```diff
@@ -194,3 +194,24 @@ def tuple(self):
         if n == 1:
             return get_point(0)
         return tuple(get_point(i) for i in range(n))
+
+    @property
+    def is_ccw(self):
+        """
+        Return whether or not this coordinate sequence has counterclockwise
+        orientation.
+        """
+        if geos_version_tuple() < (3, 7):
+            # A modified shoelace algorithm to determine polygon orientation.
+            # See https://en.wikipedia.org/wiki/Shoelace_formula.
+            area = 0.0
+            n = len(self)
+            for i in range(n):
+                j = (i + 1) % n
+                area += self[i][0] * self[j][1]
+                area -= self[j][0] * self[i][1]
+            return area > 0.0
```

Comment:
Hmm. This is a bit of an interesting one...

Should `0.0` be considered clockwise or counterclockwise? Strictly it is neither.

The removed implementation of `_isClockwise()` had `area < 0.0`, so I think this should be `area >= 0.0` which would keep the behaviour backwards compatible. That said, we should probably follow the behaviour as implemented in GEOS. Can we have a test that explicitly handles where `area == 0.0` and ensures that the behaviour matches in both cases?

Reply:
As far as I'm concerned [GEOS returns](https://github.com/libgeos/geos/blob/26c0b56ee17672617a4eeec4635ef644c33380f8/tests/unit/capi/GEOSCoordSeqTest.cpp#L335-L345) `False` for an empty sequence, so `area > 0.0` is consistent with the current GEOS behavior. @sir-sigurd Am I right?

Reply:
@felixxm, no :slightly_smiling_face: 
Actually `GEOSCoordSeq_isCCW` returns error code for empty sequences. Also GEOS and shoelace algorithm produce inconsistent results for self-intersecting `LinearRing`s, which as I understand from reading [Wikipedia](https://en.wikipedia.org/wiki/Curve_orientation) are non-orientable. So probably it's better to expose only GEOS implementation.

Reply:
Thanks for explanation.